ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound.hs

```diff
@@ -0,0 +1,527 @@
+  -- objects out of order but must use the original order when adding
+  -- to the objectPool or acknowledging objects.
+  --
+  -- However, it's worth noting that, in a few situations, some of the
+  -- object IDs in this 'Map' may be mapped to 'Nothing':
+  --
+  -- * Transaction IDs mapped to 'Nothing' can represent object IDs
+  --   that were requested, but not received. This can occur because the
+  --   client will not necessarily send all of the objects that we
+  --   asked for, but we still need to acknowledge those objects.
+  --
+  --   For example, if we request a object that no longer exists in
+  --   the client's objectPool, the client will just exclude it from the
+  --   response. However, we still need to acknowledge it (i.e. remove it
+  --   from the 'unacknowledgedObjectIds') in order to note that we're no
+  --   longer awaiting receipt of that object.
+  --
+  -- * Transaction IDs mapped to 'Nothing' can represent objects
+  --   that were not requested from the client because they're already
+  --   in the objectPool.
+  --
+  --   For example, if we request some object IDs and notice that
+  --   some subset of them have are already in the objectPool, we wouldn't
+  --   want to bother asking for those specific objects. Therefore,
+  --   we would just insert those object IDs mapped to 'Nothing' to
+  --   the 'bufferedObjects' such that those objects are acknowledged,
+  --   but never actually requested.
+  , numObjectsToAcknowledge :: !Word16
+  -- ^ The number of objects we can acknowledge on our next request
+  -- for more objects. The number here have already been removed from
+  -- 'unacknowledgedObjectIds'.
+  }
+  deriving (Show, Generic)
+
+instance
+  ( NoThunks objectId
+  , NoThunks object
+  ) =>
+  NoThunks (InboundState objectId object)
+
+initialInboundState :: InboundState objectId object
+initialInboundState = InboundState 0 Seq.empty Map.empty Map.empty 0
+
+objectDiffusionInbound ::
+  forall objectId object index m.
+  ( Ord objectId
+  , NoThunks objectId
+  , NoThunks object
+  , MonadSTM m
+  , MonadThrow m
+  ) =>
+  Tracer m (TraceObjectDiffusionInbound objectId object) ->
+  -- | Maximum number of unacknowledged objectIds allowed
+  NumObjectIdsToAck ->
+  ObjectPoolReader objectId object index m ->
+  ObjectPoolWriter objectId object m ->
+  NodeToNodeVersion ->
+  ObjectDiffusionInboundPipelined objectId object m ()
+objectDiffusionInbound tracer (NumObjectIdsToAck maxUnacked) ObjectPoolReader{objectPoolGetSnapshot} ObjectPoolWriter{wrGetObjectId, objectPoolAddObjects} _version =
+  ObjectDiffusionInboundPipelined $ do
+    continueWithStateM (inboundIdle Zero) initialInboundState
+ where
+  -- TODO #1656: replace these fixed limits by policies based on
+  -- SizeInBytes and delta-Q and the bandwidth/delay product.
+  -- These numbers are for demo purposes only, the throughput will be low.
+  maxObjectIdsToRequest = 3 :: Word16
+  maxObjectToRequest = 2 :: Word16
```

Comment:
These should be parameters given to the implementation. Probably makes sense to bundle them up with `NumObjectIdsToAck`.